README.md
# Based on (https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor)

[![hacs_badge](https://img.shields.io/badge/HACS-Custom-orange.svg)](https://hacs.xyz/docs/faq/custom_repositories)
[![Community Forum](https://img.shields.io/badge/Community-Forum-41BDF5.svg?style=popout)](https://community.home-assistant.io/t/xiaomi-cloud-vacuum-map-extractor/231292)

# Xiaomi Cloud Map Extractor

This custom integration provides a way to present a live view of a map for a Xiaomi vacuum.

<img src="https://raw.githubusercontent.com/Teka101/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor/master/images/map_no_rooms.png" width=48%>  <img src="https://raw.githubusercontent.com/Teka101/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor/master/images/map_rooms.png" width=48%>

## Installation

### Using [HACS](https://hacs.xyz/) (recommended)

This integration can be added to HACS as a [custom repository](https://hacs.xyz/docs/faq/custom_repositories):
* URL: `https://github.com/Teka101/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor`
* Category: `Integration`

After adding a custom repository you can use HACS to install this integration using user interface.

## Configuration

After installation of the custom component, it needs to be configured in `configuration.yaml` file.
To do so, add a camera entry to your configuration with at least a [basic](#basic) or [recommended](#recommended) configuration.
Vacuum token can be extracted by following [this guide](https://www.home-assistant.io/integrations/xiaomi_miio/#retrieving-the-access-token).
You also need to enter your Xiaomi Cloud username and password.
These are the credentials used for the Xiaomi Home app (_not ones from Roborock app_).

After installation and a reboot of your Home Assistant instance, you should get a camera entity which shows the vacuum map.
This might take a few minutes after a first restart.
If you have a problem with configuration validation you have to remove camera from `configuration.yaml`, restart Home Assistant, add camera config and restart HA again.

#### Available configuration parameters

| Key | Type | Required | Value | Description |
|---|---|---|---|---|
| `platform` | string | true | `xiaomi_cloud_map_extractor_v2` | Name of a platform |
| `host` | string | true | `192.168.0.123` | IP address of a vacuum |
| `token` | string | true | `ghjhca3ykg8o2zyyj7xb5adamhgsypel` | Token of a vacuum |
| `username` | string | true | `xiaomi.account@gmail.com` | Username (email or user ID) used to connect to Xiaomi cloud (the account used in the Xiaomi Home app) |
| `password` | string | true | `aVerySecretPassword` | Password used to connect to Xiaomi cloud (the account used in the Xiaomi Home app) |
| `name` | string | false |   | Desired name of camera entity |
| `country` | string | false | One of: `cn`, `de`, `us`, `ru`, `tw`, `sg`, `in`, `i2` | Server used in Xiaomi cloud. Leave empty if you are not sure. |
| `colors` | map | false |  | Colors configuration ([see below](#colors-configuration)) |
| `room_colors` | map | false |  | Room colors configuration ([see below](#room-colors-configuration)) |
| `draw` | list | false |  | List of elements to draw on a map ([see below](#draw-configuration)) |
| `texts` | list | false |  | List of texts to draw on a map ([see below](#texts-configuration)) |
| `map_transformation` | map | false |  | Parameters of map transformation ([see below](#map-transformation-configuration)) |
| `sizes` | map | false |  | Sizes of map's elements ([see below](#sizes-configuration)) |
| `attributes` | list | false |  | List of desired entity attributes ([see below](#attributes-configuration)) |
| `scan_interval` | interval | false | default: `5` seconds | Interval between map updates ([documentation](https://www.home-assistant.io/docs/configuration/platform_options/#scan-interval)) |
| `auto_update` | boolean | false | default: `true` | Activation/deactivation of automatic map updates. If disabled use service `homeassistant.update_entity` to update map manually. |

#### Colors configuration

  Each color is represented by a list of 3 or 4 parameters: `[red, green, blue]` or `[red, green, blue, alpha]`.
  Each parameter is a number from a range 0-255 and can be also provided as a HEX value: [0x12, 0xAF, 0xC5] matches #12AFC5. 
  
  <img src="https://raw.githubusercontent.com/Teka101/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor/master/images/map_no_rooms_custom_colors.png" width=50%>
   
  | Color name | Description |
  | --- | --- |
  | `color_map_inside` | Map inside (for software without rooms support) |
  | `color_map_outside` | Map outside |
  | `color_map_wall` | Walls (for software without rooms support) |
  | `color_map_wall_v2` | Walls (for software with rooms support) |
  | `color_grey_wall` | Obstacles (e.g. chairs, table legs) |
  | `color_path` | Path of a vacuum |
  | `color_goto_path` | Path for goto mode |
  | `color_predicted_path` | Predicted path to a point in goto mode |
  | `color_zones` | Fill of areas selected for zoned cleaning |
  | `color_zones_outline` | Outline of areas selected for zoned cleaning |
  | `color_virtual_walls` | Virtual walls |
  | `color_no_go_zones` | Fill of no-go zones |
  | `color_no_go_zones_outline` | Outline of no-go zones |
  | `color_no_mop_zones` | Fill of no-mopping zones |
  | `color_no_mop_zones_outline` | Outline of no-mopping zones |
  | `color_charger` | Charger position |
  | `color_obstacle` | Obstacle position |
  | `color_robo` | Vacuum position |
  | `color_scan` | Areas not assigned to any room (for software with rooms support) |
  | `color_unknown` | Other areas |

#### Room colors configuration

  This section contains mapping between room numbers and colors.
  Each color is represented by a list of 3 or 4 parameters: `[red, green, blue]` or `[red, green, blue, alpha]`.
  Each parameter is a number from a range 0-255 and can be also provided as a HEX value: [0x12, 0xAF, 0xC5] matches #12AFC5. 

#### Draw configuration

  A list of features to be drawn on a map. If all features should be drawn it can be replaced with:
  ```yaml
    draw: ["all"]
  ```
  Available values:
  - `charger`
  - `path`
  - `goto_path`
  - `predicted_path`
  - `no_go_zones`
  - `no_mopping_zones`
  - `vacuum_position`
  - `virtual_walls`
  - `zones`

#### Texts configuration

Each list entry must obey a following schema.
You can get a list of available fonts by executing this command:
```bash
fc-list | grep ttf | sed "s/.*\///"| sed "s/ttf.*/ttf/"
```

  | Parameter | Type | Required | Default value | Description |
  |---|---|---|---|---|
  | `text` | string | true |   | Text to draw on a map |
  | `x` | float | true |   | X position of a text (in percents) |
  | `y` | float | true |   | Y position of a text (in percents) |
  | `color` | list | false | black | Desired color of a text, formatted like [here](#colors-configuration) |
  | `font` | string | false |   | Name of a font to use |
  | `font_size` | int | false |   | Size of a font |

#### Map transformation configuration

  | Parameter | Type | Required | Default value | Description |
  |---|---|---|---|---|
  | `scale` | float | false | 1 | Scaling factor for a map. |
  | `rotate` | integer | false | 0 | Angle of map rotation. Available values: [`0`, `90`, `180`, `270`] |
  | `trim` | map | false | 0 | Map trimming configuration. Each trimming direction is in percents: value `25` means trimming of quarter in a given dimension. Available keys: [`left`, `right`, `top`, `bottom`] |

#### Sizes configuration

  | Parameter | Type | Required | Default value | Description |
  |---|---|---|---|---|
  | `charger_radius` | float | false | 4 | Radius of a charger circle. |
  | `vacuum_radius` | float | false | 4 | Radius of a vacuum circle. |

#### Attributes configuration

  A list of attributes that an entity should have.
  Available values:
  - `calibration_points` - Calculated calibration points for [Lovelace Xiaomi Vacuum Map card](https://github.com/PiotrMachowski/lovelace-xiaomi-vacuum-map-card).

     <img src="https://raw.githubusercontent.com/Teka101/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor/master/images/map_card.gif" width=50%>
  
  - `charger`
  - `country`
  - `goto`
  - `goto_path`
  - `goto_predicted_path`
  - `image`
  - `is_empty`
  - `map_name`
  - `no_go_areas`
  - `no_mopping_areas`
  - `obstacles`
  - `path`
  - `room_numbers`
  - `rooms`
  - `vacuum_position`
  - `vacuum_room`
  - `walls`
  - `zones`
  
### Examples

#### Basic

```yaml
camera:
  - platform: xiaomi_cloud_map_extractor_v2
    host: !secret xiaomi_vacuum_host
    token: !secret xiaomi_vacuum_token
    username: !secret xiaomi_cloud_username
    password: !secret xiaomi_cloud_password
```

#### Recommended

```yaml
camera:
  - platform: xiaomi_cloud_map_extractor_v2
    host: !secret xiaomi_vacuum_host
    token: !secret xiaomi_vacuum_token
    username: !secret xiaomi_cloud_username
    password: !secret xiaomi_cloud_password
    draw: ['all']
    attributes:
      - calibration_points    
```


#### Full

```yaml
camera:
  - platform: xiaomi_cloud_map_extractor_v2
    host: !secret xiaomi_vacuum_host
    token: !secret xiaomi_vacuum_token
    username: !secret xiaomi_cloud_username
    password: !secret xiaomi_cloud_password
    country: "de"
    name: "My Vacuum Camera"
    colors:
      color_map_inside: [32, 115, 185]
      color_map_outside: [19, 87, 148]
      color_map_wall: [100, 196, 254]
      color_map_wall_v2: [93, 109, 126]
      color_grey_wall: [93, 109, 126]
      color_path: [147, 194, 238]
      color_goto_path: [0, 255, 0]
      color_predicted_path: [255, 255, 0, 0]
      color_zones: [0xAD, 0xD8, 0xFF, 0x8F]
      color_zones_outline: [0xAD, 0xD8, 0xFF]
      color_virtual_walls: [255, 0, 0]
      color_obstacle: [0, 0, 0, 180]
      color_no_go_zones: [255, 33, 55, 127]
      color_no_go_zones_outline: [255, 0, 0]
      color_no_mop_zones: [163, 130, 211, 127]
      color_no_mop_zones_outline: [163, 130, 211]
      color_charger: [0x66, 0xfe, 0xda, 0x7f]
      color_robo: [75, 235, 149]
      color_unknown: [0, 0, 0]
      color_scan: [0xDF, 0xDF, 0xDF]
    room_colors:
      1: [240, 178, 122]
      2: [133, 193, 233]
      3: [217, 136, 128]
      4: [52, 152, 219]
      5: [205, 97, 85]
      6: [243, 156, 18]
      7: [88, 214, 141]
      8: [245, 176, 65]
      9: [252, 212, 81]
      10: [72, 201, 176]
      11: [84, 153, 199]
      12: [133, 193, 233]
      13: [245, 176, 65]
      14: [82, 190, 128]
      15: [72, 201, 176]
      16: [165, 105, 18]
    draw:
      - charger
      - path
      - goto_path
      - predicted_path
      - obstacles
      - no_go_zones
      - no_mopping_zones
      - vacuum_position
      - virtual_walls
      - zones
    texts:
      - text: "Room 1"
        x: 25
        y: 25
        color: [125, 20, 213]
      - text: "Room 2"
        x: 25
        y: 75
        color: [125, 20, 213, 127]
        font: "FreeSans.ttf"
        font_size: 25
    map_transformation:
      scale: 2
      rotate: 180
      trim:
        top: 10
        bottom: 20
        left: 30
        right: 40
    sizes:
      charger_radius: 4
      obstacle_radius: 3
      vacuum_radius: 6.5
    attributes:
      - calibration_points
      - charger
      - country
      - goto
      - goto_path
      - goto_predicted_path
      - image
      - is_empty
      - map_name
      - no_go_areas
      - no_mopping_areas
      - obstacles
      - path
      - room_numbers
      - rooms
      - vacuum_position
      - vacuum_room
      - walls
      - zones
    scan_interval:
      seconds: 10
    auto_update: true
```

## Supported devices

This integration was tested on following vacuums:
 - Xiaomi Vacuum Gen 1 (Mi Robot Vacuum/SDJQR01RR/SDJQR02RR)
 - Xiaomi Mi Robot 1S 
 - Roborock S4 (software with rooms support)
 - Roborock S5 (software without rooms support)
 - Roborock S5 (software with rooms support)
 - Roborock S5 Max
 - Roborock S6
 - Roborock S6 MaxV
 - Roborock S6 Pure
 
## Unsupported devices

At this moment this integration is known to not work with following vacuums:
 - Roborock E50
 - Xiaomi Mi Robot Vacuum Mop Pro (STYJ02YM)


CHANGELOG.md
# Changelog

## [Unreleased] / master
- add drawing of detected obstacles
- improve drawing when scaling (smoother lines)

custom_components/xiaomi_cloud_map_extractor/const.py
CONF_ATTRIBUTES = "attributes"
CONF_AUTO_UPDATE = "auto_update"
CONF_AVAILABLE_COUNTRIES = ["cn", "de", "us", "ru", "tw", "sg", "in", "i2"]
CONF_BOTTOM = "bottom"
CONF_COLOR = "color"
CONF_COLORS = "colors"
CONF_COUNTRY = "country"
CONF_DRAW = "draw"
CONF_FONT = "font"
CONF_FONT_SIZE = "font_size"
CONF_LEFT = "left"
CONF_MAP_TRANSFORM = "map_transformation"
CONF_RIGHT = "right"
CONF_ROOM_COLORS = "room_colors"
CONF_ROTATE = "rotate"
CONF_SCALE = "scale"
CONF_SIZES = "sizes"
CONF_SIZE_CHARGER_RADIUS = "charger_radius"
CONF_SIZE_OBSTACLE_RADIUS = "obstacle_radius"
CONF_SIZE_VACUUM_RADIUS = "vacuum_radius"
CONF_TEXT = "text"
CONF_TEXTS = "texts"
CONF_TOP = "top"
CONF_TRIM = "trim"
CONF_X = "x"
CONF_Y = "y"

MINIMAL_IMAGE_WIDTH = 20
MINIMAL_IMAGE_HEIGHT = 20
CONTENT_TYPE = "image/png"
DEFAULT_NAME = "Xiaomi Cloud Map Extractor"

ATTRIBUTE_CALIBRATION = "calibration_points"
ATTRIBUTE_CHARGER = "charger"
ATTRIBUTE_COUNTRY = "country"
ATTRIBUTE_GOTO = "goto"
ATTRIBUTE_GOTO_PATH = "goto_path"
ATTRIBUTE_GOTO_PREDICTED_PATH = "goto_predicted_path"
ATTRIBUTE_IMAGE = "image"
ATTRIBUTE_IS_EMPTY = "is_empty"
ATTRIBUTE_MAP_NAME = "map_name"
ATTRIBUTE_NO_GO_AREAS = "no_go_areas"
ATTRIBUTE_NO_MOPPING_AREAS = "no_mopping_areas"
ATTRIBUTE_OBSTACLES = "obstacles"
ATTRIBUTE_PATH = "path"
ATTRIBUTE_ROOM_NUMBERS = "room_numbers"
ATTRIBUTE_ROOMS = "rooms"
ATTRIBUTE_VACUUM_POSITION = "vacuum_position"
ATTRIBUTE_VACUUM_ROOM = "vacuum_room"
ATTRIBUTE_WALLS = "walls"
ATTRIBUTE_ZONES = "zones"

CONF_AVAILABLE_ATTRIBUTES = [ATTRIBUTE_CALIBRATION, ATTRIBUTE_CHARGER, ATTRIBUTE_COUNTRY, ATTRIBUTE_GOTO,
                             ATTRIBUTE_GOTO_PATH, ATTRIBUTE_GOTO_PREDICTED_PATH, ATTRIBUTE_IMAGE, ATTRIBUTE_IS_EMPTY,
                             ATTRIBUTE_MAP_NAME, ATTRIBUTE_NO_GO_AREAS, ATTRIBUTE_NO_MOPPING_AREAS, ATTRIBUTE_OBSTACLES,
                             ATTRIBUTE_PATH, ATTRIBUTE_ROOM_NUMBERS, ATTRIBUTE_ROOMS, ATTRIBUTE_VACUUM_POSITION,
                             ATTRIBUTE_VACUUM_ROOM, ATTRIBUTE_WALLS, ATTRIBUTE_ZONES]

ATTR_A = "a"
ATTR_ANGLE = "angle"
ATTR_HEIGHT = "height"
ATTR_OFFSET_X = "offset_x"
ATTR_OFFSET_Y = "offset_y"
ATTR_PATH = "path"
ATTR_POINT_LENGTH = "point_length"
ATTR_POINT_SIZE = "point_size"
ATTR_ROTATION = "rotation"
ATTR_SCALE = "scale"
ATTR_SIZE = "size"
ATTR_WIDTH = "width"
ATTR_X = "x"
ATTR_X0 = "x0"
ATTR_X1 = "x1"
ATTR_X2 = "x2"
ATTR_X3 = "x3"
ATTR_Y = "y"
ATTR_Y0 = "y0"
ATTR_Y1 = "y1"
ATTR_Y2 = "y2"
ATTR_Y3 = "y3"

DRAWABLE_ALL = "all"
DRAWABLE_CHARGER = "charger"
DRAWABLE_GOTO_PATH = "goto_path"
DRAWABLE_OBSTACLES = "obstacles"
DRAWABLE_NO_GO_AREAS = "no_go_zones"
DRAWABLE_NO_MOPPING_AREAS = "no_mopping_zones"
DRAWABLE_PATH = "path"
DRAWABLE_PREDICTED_PATH = "predicted_path"
DRAWABLE_VACUUM_POSITION = "vacuum_position"
DRAWABLE_VIRTUAL_WALLS = "virtual_walls"
DRAWABLE_ZONES = "zones"

CONF_AVAILABLE_DRAWABLES = [DRAWABLE_ALL, DRAWABLE_NO_MOPPING_AREAS, DRAWABLE_NO_GO_AREAS, DRAWABLE_VIRTUAL_WALLS,
                            DRAWABLE_CHARGER, DRAWABLE_ZONES, DRAWABLE_OBSTACLES, DRAWABLE_GOTO_PATH, DRAWABLE_PATH,
                            DRAWABLE_PREDICTED_PATH, DRAWABLE_VACUUM_POSITION]

COLOR_ROOM_PREFIX = "color_room_"

COLOR_CHARGER = "color_charger"
COLOR_GOTO_PATH = "color_goto_path"
COLOR_GREY_WALL = "color_grey_wall"
COLOR_MAP_INSIDE = "color_map_inside"
COLOR_MAP_OUTSIDE = "color_map_outside"
COLOR_MAP_WALL = "color_map_wall"
COLOR_MAP_WALL_V2 = "color_map_wall_v2"
COLOR_NO_GO_ZONES = "color_no_go_zones"
COLOR_NO_GO_ZONES_OUTLINE = "color_no_go_zones_outline"
COLOR_NO_MOPPING_ZONES = "color_no_mop_zones"
COLOR_NO_MOPPING_ZONES_OUTLINE = "color_no_mop_zones_outline"
COLOR_OBSTACLE = "color_obstacle"
COLOR_PATH = "color_path"
COLOR_PREDICTED_PATH = "color_predicted_path"
COLOR_ROBO = "color_robo"
COLOR_SCAN = "color_scan"
COLOR_UNKNOWN = "color_unknown"
COLOR_VIRTUAL_WALLS = "color_virtual_walls"
COLOR_ZONES = "color_zones"
COLOR_ZONES_OUTLINE = "color_zones_outline"

CONF_AVAILABLE_COLORS = [COLOR_CHARGER, COLOR_GOTO_PATH, COLOR_GREY_WALL, COLOR_MAP_INSIDE, COLOR_MAP_OUTSIDE,
                         COLOR_MAP_WALL, COLOR_MAP_WALL_V2, COLOR_NO_GO_ZONES, COLOR_NO_GO_ZONES_OUTLINE,
                         COLOR_NO_MOPPING_ZONES, COLOR_NO_MOPPING_ZONES_OUTLINE, COLOR_PATH, COLOR_PREDICTED_PATH,
                         COLOR_ROBO, COLOR_SCAN, COLOR_UNKNOWN, COLOR_VIRTUAL_WALLS, COLOR_ZONES, COLOR_ZONES_OUTLINE,
                         COLOR_OBSTACLE]

COLOR_ROOM_1 = "color_room_1"
COLOR_ROOM_2 = "color_room_2"
COLOR_ROOM_3 = "color_room_3"
COLOR_ROOM_4 = "color_room_4"
COLOR_ROOM_5 = "color_room_5"
COLOR_ROOM_6 = "color_room_6"
COLOR_ROOM_7 = "color_room_7"
COLOR_ROOM_8 = "color_room_8"
COLOR_ROOM_9 = "color_room_9"
COLOR_ROOM_10 = "color_room_10"
COLOR_ROOM_11 = "color_room_11"
COLOR_ROOM_12 = "color_room_12"
COLOR_ROOM_13 = "color_room_13"
COLOR_ROOM_14 = "color_room_14"
COLOR_ROOM_15 = "color_room_15"
COLOR_ROOM_16 = "color_room_16"

CONF_DEFAULT_ROOM_COLORS = [COLOR_ROOM_1, COLOR_ROOM_2, COLOR_ROOM_3, COLOR_ROOM_4, COLOR_ROOM_5, COLOR_ROOM_6,
                            COLOR_ROOM_7, COLOR_ROOM_8, COLOR_ROOM_9, COLOR_ROOM_10, COLOR_ROOM_11, COLOR_ROOM_12,
                            COLOR_ROOM_13, COLOR_ROOM_14, COLOR_ROOM_15, COLOR_ROOM_16]


custom_components/xiaomi_cloud_map_extractor/image_handler.py
import logging
from typing import Callable
from PIL import Image, ImageDraw, ImageFont

from .const import *

_LOGGER = logging.getLogger(__name__)


class ImageHandler:
    MAP_OUTSIDE = 0x00
    MAP_WALL = 0x01
    MAP_INSIDE = 0xFF
    MAP_SCAN = 0x07
    COLORS = {
        COLOR_MAP_INSIDE: (32, 115, 185),
        COLOR_MAP_OUTSIDE: (19, 87, 148),
        COLOR_MAP_WALL: (100, 196, 254),
        COLOR_MAP_WALL_V2: (93, 109, 126),
        COLOR_GREY_WALL: (93, 109, 126),
        COLOR_PATH: (147, 194, 238),
        COLOR_GOTO_PATH: (0, 255, 0),
        COLOR_PREDICTED_PATH: (255, 255, 0),
        COLOR_ZONES: (0xAD, 0xD8, 0xFF, 0x8F),
        COLOR_ZONES_OUTLINE: (0xAD, 0xD8, 0xFF),
        COLOR_VIRTUAL_WALLS: (255, 0, 0),
        COLOR_NO_GO_ZONES: (255, 33, 55, 127),
        COLOR_NO_GO_ZONES_OUTLINE: (255, 0, 0),
        COLOR_NO_MOPPING_ZONES: (163, 130, 211, 127),
        COLOR_NO_MOPPING_ZONES_OUTLINE: (163, 130, 211),
        COLOR_CHARGER: (0x66, 0xfe, 0xda, 0x7f),
        COLOR_ROBO: (75, 235, 149),
        COLOR_UNKNOWN: (0, 0, 0),
        COLOR_SCAN: (0xDF, 0xDF, 0xDF),
        COLOR_ROOM_1: (240, 178, 122),
        COLOR_ROOM_2: (133, 193, 233),
        COLOR_ROOM_3: (217, 136, 128),
        COLOR_ROOM_4: (52, 152, 219),
        COLOR_ROOM_5: (205, 97, 85),
        COLOR_ROOM_6: (243, 156, 18),
        COLOR_ROOM_7: (88, 214, 141),
        COLOR_ROOM_8: (245, 176, 65),
        COLOR_ROOM_9: (252, 212, 81),
        COLOR_ROOM_10: (72, 201, 176),
        COLOR_ROOM_11: (84, 153, 199),
        COLOR_ROOM_12: (133, 193, 233),
        COLOR_ROOM_13: (245, 176, 65),
        COLOR_ROOM_14: (82, 190, 128),
        COLOR_ROOM_15: (72, 201, 176),
        COLOR_ROOM_16: (165, 105, 189)
    }
    ROOM_COLORS = [COLOR_ROOM_1, COLOR_ROOM_2, COLOR_ROOM_3, COLOR_ROOM_4, COLOR_ROOM_5, COLOR_ROOM_6, COLOR_ROOM_7,
                   COLOR_ROOM_8, COLOR_ROOM_9, COLOR_ROOM_10, COLOR_ROOM_11, COLOR_ROOM_12, COLOR_ROOM_13,
                   COLOR_ROOM_14, COLOR_ROOM_15, COLOR_ROOM_16]

    @staticmethod
    def parse(raw_data: bytes, width, height, colors, image_config):
        rooms = {}
        scale = image_config[CONF_SCALE]
        trim_left = int(image_config[CONF_TRIM][CONF_LEFT] * width / 100)
        trim_right = int(image_config[CONF_TRIM][CONF_RIGHT] * width / 100)
        trim_top = int(image_config[CONF_TRIM][CONF_TOP] * height / 100)
        trim_bottom = int(image_config[CONF_TRIM][CONF_BOTTOM] * height / 100)
        trimmed_height = height - trim_top - trim_bottom
        trimmed_width = width - trim_left - trim_right
        image = Image.new('RGBA', (trimmed_width, trimmed_height))
        if width == 0 or height == 0:
            return ImageHandler.create_empty_map(colors)
        pixels = image.load()
        for img_y in range(trimmed_height):
            for img_x in range(trimmed_width):
                pixel_type = raw_data[img_x + trim_left + width * (img_y + trim_bottom)]
                x = img_x
                y = trimmed_height - img_y - 1
                if pixel_type == ImageHandler.MAP_OUTSIDE:
                    pixels[x, y] = ImageHandler.__get_color__(COLOR_MAP_OUTSIDE, colors)
                elif pixel_type == ImageHandler.MAP_WALL:
                    pixels[x, y] = ImageHandler.__get_color__(COLOR_MAP_WALL, colors)
                elif pixel_type == ImageHandler.MAP_INSIDE:
                    pixels[x, y] = ImageHandler.__get_color__(COLOR_MAP_INSIDE, colors)
                elif pixel_type == ImageHandler.MAP_SCAN:
                    pixels[x, y] = ImageHandler.__get_color__(COLOR_SCAN, colors)
                else:
                    obstacle = pixel_type & 0x07
                    if obstacle == 0:
                        pixels[x, y] = ImageHandler.__get_color__(COLOR_GREY_WALL, colors)
                    elif obstacle == 1:
                        pixels[x, y] = ImageHandler.__get_color__(COLOR_MAP_WALL_V2, colors)
                    elif obstacle == 7:
                        room_number = (pixel_type & 0xFF) >> 3
                        room_x = img_x + trim_left
                        room_y = img_y + trim_bottom
                        if room_number not in rooms:
                            rooms[room_number] = (room_x, room_y, room_x, room_y)
                        else:
                            rooms[room_number] = (min(rooms[room_number][0], room_x),
                                                  min(rooms[room_number][1], room_y),
                                                  max(rooms[room_number][2], room_x),
                                                  max(rooms[room_number][3], room_y))
                        default = ImageHandler.ROOM_COLORS[room_number >> 1]
                        pixels[x, y] = ImageHandler.__get_color__(f"{COLOR_ROOM_PREFIX}{room_number}", colors, default)
                    else:
                        pixels[x, y] = ImageHandler.__get_color__(COLOR_UNKNOWN, colors)
        if image_config["scale"] != 1 and width != 0 and height != 0:
            image = image.resize((int(trimmed_width * scale), int(trimmed_height * scale)), resample=Image.NEAREST)
        return image, rooms

    @staticmethod
    def create_empty_map(colors):
        color = ImageHandler.__get_color__(COLOR_MAP_OUTSIDE, colors)
        image = Image.new('RGBA', (100, 100), color=color)
        if sum(color[0:3]) > 382:
            text_color = (0, 0, 0)
        else:
            text_color = (255, 255, 255)
        draw = ImageDraw.Draw(image, "RGBA")
        text = "NO MAP"
        w, h = draw.textsize(text)
        draw.text((50 - w / 2, 50 - h / 2), text, fill=text_color)
        return image, {}

    @staticmethod
    def get_room_at_pixel(raw_data: bytes, width, x, y):
        room_number = None
        pixel_type = raw_data[x + width * y]
        if pixel_type not in [ImageHandler.MAP_INSIDE, ImageHandler.MAP_SCAN]:
            if pixel_type & 0x07 == 7:
                room_number = (pixel_type & 0xFF) >> 3
        return room_number

    @staticmethod
    def draw_path(image, path, colors, scale):
        ImageHandler.__draw_path__(image, path, ImageHandler.__get_color__(COLOR_PATH, colors), scale)

    @staticmethod
    def draw_goto_path(image, path, colors, scale):
        ImageHandler.__draw_path__(image, path, ImageHandler.__get_color__(COLOR_GOTO_PATH, colors), scale)

    @staticmethod
    def draw_predicted_path(image, path, colors, scale):
        ImageHandler.__draw_path__(image, path, ImageHandler.__get_color__(COLOR_PREDICTED_PATH, colors), scale)

    @staticmethod
    def draw_no_go_areas(image, areas, colors):
        ImageHandler.__draw_areas__(image, areas,
                                    ImageHandler.__get_color__(COLOR_NO_GO_ZONES, colors),
                                    ImageHandler.__get_color__(COLOR_NO_GO_ZONES_OUTLINE, colors))

    @staticmethod
    def draw_no_mopping_areas(image, areas, colors):
        ImageHandler.__draw_areas__(image, areas,
                                    ImageHandler.__get_color__(COLOR_NO_MOPPING_ZONES, colors),
                                    ImageHandler.__get_color__(COLOR_NO_MOPPING_ZONES_OUTLINE, colors))

    @staticmethod
    def draw_walls(image, walls, colors):
        draw = ImageDraw.Draw(image.data, 'RGBA')
        for wall in walls:
            draw.line(wall.to_img(image.dimensions).as_list(),
                      ImageHandler.__get_color__(COLOR_VIRTUAL_WALLS, colors), width=2)

    @staticmethod
    def draw_zones(image, zones, colors):
        areas = list(map(lambda z: z.as_area(), zones))
        ImageHandler.__draw_areas__(image, areas,
                                    ImageHandler.__get_color__(COLOR_ZONES, colors),
                                    ImageHandler.__get_color__(COLOR_ZONES_OUTLINE, colors))

    @staticmethod
    def draw_charger(image, charger, radius, colors):
        color = ImageHandler.__get_color__(COLOR_CHARGER, colors)
        ImageHandler.__draw_circle__(image, charger, radius, color, color)

    @staticmethod
    def draw_obstacles(image, obstacles, radius, colors):
        color = ImageHandler.__get_color__(COLOR_OBSTACLE, colors)
        for obstacle in obstacles:
            ImageHandler.__draw_circle__(image, obstacle, radius, color, color)

    @staticmethod
    def draw_vacuum_position(image, vacuum_position, radius, colors):
        color = ImageHandler.__get_color__(COLOR_ROBO, colors)
        ImageHandler.__draw_circle__(image, vacuum_position, radius, color, color)

    @staticmethod
    def rotate(image):
        if image.dimensions.rotation == 90:
            image.data = image.data.transpose(Image.ROTATE_90)
        if image.dimensions.rotation == 180:
            image.data = image.data.transpose(Image.ROTATE_180)
        if image.dimensions.rotation == 270:
            image.data = image.data.transpose(Image.ROTATE_270)

    @staticmethod
    def draw_texts(image, texts):
        for text_config in texts:
            x = text_config[CONF_X] * image.data.size[0] / 100
            y = text_config[CONF_Y] * image.data.size[1] / 100
            ImageHandler.__draw_text__(image, text_config[CONF_TEXT], x, y, text_config[CONF_COLOR],
                                       text_config[CONF_FONT], text_config[CONF_FONT_SIZE])

    @staticmethod
    def __draw_circle__(image, center, r, outline, fill):
        def draw_func(draw: ImageDraw):
            point = center.to_img(image.dimensions)
            coords = [point.x - r, point.y - r, point.x + r, point.y + r]
            draw.ellipse(coords, outline=outline, fill=fill)

        ImageHandler.__draw_on_new_layer__(image, draw_func)

    @staticmethod
    def __draw_areas__(image, areas, fill, outline):
        if len(areas) == 0:
            return

        def draw_func(draw: ImageDraw):
            for area in areas:
                draw.polygon(area.to_img(image.dimensions).as_list(), fill, outline)

        ImageHandler.__draw_on_new_layer__(image, draw_func)

    @staticmethod
    def __draw_path__(image, path, color, scale):
        if len(path.path) < 2:
            return

        def draw_func(draw: ImageDraw):
            s = path.path[0].to_img(image.dimensions)
            for point in path.path[1:]:
                e = point.to_img(image.dimensions)
                draw.line([s.x * scale, s.y * scale, e.x * scale, e.y * scale], width=scale, fill=color)
                s = e

        ImageHandler.__draw_on_new_layer__(image, draw_func, scale)

    @staticmethod
    def __draw_text__(image, text, x, y, color, font_file=None, font_size=None):
        def draw_func(draw: ImageDraw):
            font = ImageFont.load_default()
            try:
                if font_file is not None and font_size > 0:
                    font = ImageFont.truetype(font_file, font_size)
            except OSError:
                _LOGGER.warning("Unable to find font file: %s", font_file)
            except ImportError:
                _LOGGER.warning("Unable to open font: %s", font_file)
            finally:
                w, h = draw.textsize(text, font)
                draw.text((x - w / 2, y - h / 2), text, font=font, fill=color)

        ImageHandler.__draw_on_new_layer__(image, draw_func)

    @staticmethod
    def __get_color__(name, colors, default_name=None):
        if name in colors:
            return colors[name]
        if default_name is None:
            return ImageHandler.COLORS[name]
        return ImageHandler.COLORS[default_name]

    @staticmethod
    def __draw_on_new_layer__(image, draw_function: Callable, scale=1):
        if scale == 1:
            size = image.data.size
        else:
            size = [image.data.size[0] * scale, image.data.size[1] * scale]
        layer = Image.new("RGBA", size, (255, 255, 255, 0))
        draw = ImageDraw.Draw(layer, "RGBA")
        draw_function(draw)
        if scale > 1:
            layer = layer.resize(image.data.size, resample=Image.BOX)
        image.data = Image.alpha_composite(image.data, layer)
